port/esp32/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,828 @@
+    },
+    {
+      .key = "esp32.chipRevision",
+      .value = revision,
+    },
+    {
+      .key = "esp32.sdkVersion",
+      .value = esp_get_idf_version(),
+    },
+  };
+  const size_t data_size = sizeof(data) / sizeof(data[0]);
+
+  if (url == NULL || url[0] == '\0') {
+    APP_ERR("hawkbit configure error: invalid configuration URL");
+    return -1;
+  }
+
+  APP_DBG("hawkbit configure at href(%s)", url);
+  cJSON *root = cJSON_CreateObject();
+
+  // mode
+  cJSON_AddItemToObject(root, "mode", cJSON_CreateString("replace"));
+
+  // data
+  cJSON *jdata = cJSON_CreateObject();
+  cJSON_AddItemToObject(root, "data", jdata);
+  for (size_t i = 0; i < data_size; ++i) {
+    cJSON_AddItemToObject(jdata, data[i].key,
+                          cJSON_CreateString(data[i].value));
+  }
+
+  cJSON *status = cJSON_CreateObject();
+  cJSON_AddItemToObject(root, "status", status);
+  // status.execution
+  cJSON_AddItemToObject(status, "execution", cJSON_CreateString("closed"));
+  cJSON *status_result = cJSON_CreateObject();
+  // status.result
+  cJSON_AddItemToObject(status, "result", status_result);
+  // status.result.finished
+  cJSON_AddItemToObject(status_result, "finished",
+                        cJSON_CreateString("success"));
+
+  char *body = cJSON_PrintUnformatted(root);
+  cJSON_Delete(root);
+  if (body == NULL) {
+    APP_ERR("hawkbit configure error: cannot render JSON object");
+    return false;
+  }
+
+  char output[HAWKBIT_HTTP_MAX_OUTPUT_BUFFER] = { 0 };
+  int code = hawkbit_http_perform_put(url, body, output, sizeof(output));
+  free(body);
+  if (code < 0 || code != HAWKBIT_HTTP_CODE_OK) {
+    APP_ERR("hawkbit configure error: unexpected HTTP code(%d)", code);
+    return false;
+  }
+  return true;
+}
+
+static esp_http_client_config_t
+hawkbit_download_get_client_config(hawkbit_download_links_t links)
+{
+  if (oc_string(links.download) != NULL) {
+    // TODO
+  }
+  if (oc_string(links.downloadHttp) != NULL) {
+    esp_http_client_config_t config = {};
+    config.url = oc_string(links.downloadHttp);
+    config.cert_pem = (char *)"";
+    config.skip_cert_common_name_check = true;
+    return config;
+  }
+  esp_http_client_config_t config = {};
+  return config;
+}
+
+bool
+hawkbit_download_from_links(hawkbit_download_links_t links)
+{
+  if (oc_string(links.download) == NULL &&
+      oc_string(links.downloadHttp) == NULL) {
+    return false;
+  }
+  esp_http_client_config_t config = hawkbit_download_get_client_config(links);
+  esp_https_ota_config_t ota_config = {
+    .http_config = &config,
+    .partial_http_download = false,
+  };
+  esp_https_ota_handle_t https_ota_handle = NULL;
+  esp_err_t err = esp_https_ota_begin(&ota_config, &https_ota_handle);
+  if (err != ESP_OK) {
+    APP_ERR("OTA update begin failed: %s", esp_err_to_name(err));
+    return false;
+  }
+
+  while (true) {
+    err = esp_https_ota_perform(https_ota_handle);
+    if (err != ESP_ERR_HTTPS_OTA_IN_PROGRESS) {
```

Comment:
how othen tthis is called -> maybe here we can plan delayed callback to handing iotivity-lite events